docs/TPPDialect.md

```diff
@@ -0,0 +1,351 @@
+# TPP Dialect
+
+The TPP dialect sits between `linalg` and `xsmm` and represents the high-level TPP operations that we want to be the virtual-ISA, compatible with multiple architectures (CPUs, GPUs, etc).
+
+After the initial cleanup phase, where ingress dialects are converted to standard dialects (`linalg`, `tcp`, `scf`), we match those patterns to (a sequence of) TPP operations.
+
+Those operations can be lowered to individual `xsmm` operations (which themselves are lowered to `libxsmm` calls), but can also be fused into other TPP operations, forming _fused_ operations.
+
+The TPP dialect supports `tensor` types, converted directly from `linalg`, `tensor` and `tcp` operations, and have their own tiling and bufferization rules at `tensor` level.
+
+After bufferization, the dialect also supports `memref` types, where further grouping, transforms and annotation happen, in preparaiton for optimal `xsmm` lowering.
+
+## Optimizations
+
+Due to cache locality, instruction set availability and number of registers, there are various costs and benefits of picking particular combinations, which aren't immediately clear to the programmer.
+
+Our compiler needs to select the most appropriate combination that can be lowered to existing and optiomal `libxsmm` calls with the right blockng and tiling factors.
+
+TPP operations can carry attributes, their operands can be chained and grouped, forming trees and graphs, which can be mathed in different ways.
+
+Once lowered to `xsmm` dialect, the operations are selected and any further lowering won't change the TPP semantics, only add `xsmm` boilerplate and infrastruture to execute the chosen strategies.
+
+# Basic Rules
+
+## Syntax
+
+All `tensor` operations follow the SSA syntax, operands and return values:
+```mlir
+  // Always out-of-place
+  %C = tpp.add ins(%arg0, %arg1) : (tensor<>, tensor<>) -> tensor<>
+  // %C is here treated as initialization, accumulation can still happen on same buffer
+  %M = tpp.matmul ins(%A, %B, %C) : (tensor<>, tensor<>, tensor<>) -> tensor<>
+```
+
+All `memref` operations follow the `linalg` syntax of `ins` and `outs`.
+The `outs` values are an indication of where the output is written to.
+These are generated by the bufferization process, from the `tensor` level operations.
+```mlir
+  // Out-of-place ops output on new allocations
+  %buf = memref.alloc : memref<>
+  tpp.add ins(%arg0, %arg1) outs(%buf) : (memref<>, memref<>)
+  // In-of-place ops output on input operands
+  tpp.matmul ins(%A, %B, %C) outs(%C) : (memref<>, memref<>, memref<>)
+```
+
+### Unary
+
+On `tensors`, unary ops have one argument and one return value.
+
+On `memrefs`, unary operations have only `outs` for non-transforms (neg, inv, exp, etc.).
```

Comment:
Does it imply in-place semantics for these on memrefs?

Reply:
It does not. We discussed this a bit more and we'll common up the syntax with `ins` always present.

On `tensors`, there is no in-place. On `memrefs`, an op is in-place if the `outs` alias with at least one of the `ins`.

I'm updating this part to make that clear.